rfcs/2020-05-25-2685-dev-workflow-simplification.md

```diff
@@ -0,0 +1,244 @@
+
+### Use a Docker or Podman environment
+
+> **Targets:** You can use this method to produce AARCH64, Arm6/7, as well as x86/64 Linux builds.
+
+Since not everyone has a full working native environment, or can use Nix, we took our Nix environment and stuffed it into a Docker (or Podman) container!
+
+This is ideal for users who want it to "Just work" and just want to start contributing. It's also what we use for our CI, so you know if it breaks we can't do anything else until we fix it. 😉
+
+**Before you go farther, install Docker or Podman through your official package manager, or from the [Docker](https://docs.docker.com/get-docker/) or [Podman](https://podman.io/) sites.**
+
+```bash
+# Optional: Only if you use `podman`
+export container-tool="podman"
+```
+
+By default, `make environment` style tasks will do a `docker pull` from Github's container repository, you can **optionally** build your own environment while you make your morning coffee ☕:
+
+```bash
+# Optional: Only if you want to go make a coffee
+make environment-prepare
+```
+
+Now that you have your coffee, you can run any of the following:
+
+```bash
+# Enter a shell with optimized mounts for interactive processes.
+# Inside here, you can use Vector like you have full toolchain (See below!)
+make environment
+# Add extra cli opts
+make enviroment cli-opts="--publish 3000:2000"
+# Validate your code can compile
+make enviroment-check
+# Validate your code actually does compile (in dev mode)
+make environment-build-dev
+# Validate your test pass
+make environment-test scope="sources::example"
+# Validate tests (that do not require other services) pass
+make environment-test
+# Validate your tests pass (starting required services in Docker)
+make environment-test-integration scope="sources::example"
+# Validate your tests pass against a live service.
+make environment-test-integration scope="sources::example" autospawn=false
+# Validate all tests pass (starting required services in Docker)
+make environment-test-integration
+# Run your benchmarks
+make environment-bench scope="transforms::example"
+# Rebuild Vector's metadata
+make environment-generate
+# Serve the website on port 3000
+make environment-website
+# Format your code before pushing!
+make environment-fmt
+# Try out a specific container tool. (Docker/Podman)
+make environment container-tool="podman"
+```
+
+We use explicit environment opt-in as many contributors choose to keep their Rust toolchain local, and use `make environment-generate` etc.
+
+### Using Nix
+
+> **Note:** We're still new at this Nix stuff, so if you're a Nix expert do feel free to share knowledge.
+
+If you're a Nix user, or you're open to trying out a new tool, you can use our Nix expressions!
+
+If you don't have Nix yet, [install it](https://nixos.org/download.html):
+
+```bash
+curl -L [https://nixos.org/nix/install](https://nixos.org/nix/install) | sh
+# Hesitating? Uninstalling is just `rm -rf /nix`
```

Comment:
As an aside, and not related to this RFC (this line is a quote from the Nix home page), this isn't entirely true. The Nix installer drops some `.nix-*` files into the installing user's home directory and modifies `.bash_profile` for bash users (presumably other files for different shells). As such, uninstalling is a little more complicated, but only a little.

Reply:
Whoops! Thanks so much for catching that!